<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:arg name="namespace" default="" />
  <xacro:arg name="sim_mode" default="false" />

  <!--
  <xacro:unless value="$(arg sim_mode)">
    <ros2_control name="fake" type="system">
      <hardware>
        <plugin>fake_components/GenericSystem</plugin>
      </hardware>
      <joint name="left_wheel_joint">
        <command_interface name="velocity" />
        <state_interface name="position" />
      </joint>
      <joint name="right_wheel_joint">
        <command_interface name="velocity" />
        <state_interface name="position" />
      </joint>
    </ros2_control>
  </xacro:unless>
  -->

  <xacro:unless value="$(arg sim_mode)">
    <ros2_control name="RealRobot" type="system">
      <hardware>
        <plugin>diffdrive_arduino/DiffDriveArduinoHardware</plugin>
        <param name="left_wheel_name">left_wheel_joint</param>
        <param name="right_wheel_name">right_wheel_joint</param>
        <param name="loop_rate">30</param>
        <param name="device">/dev/ttyACM0</param>
        <param name="baud_rate">57600</param>
        <param name="timeout">1000</param>
        <param name="open_loop">true</param>
        <!--<param name="enc_counts_per_rev">1</param>-->
      </hardware>
      <joint name="left_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="position"/> 
        <state_interface name="velocity"/>
      </joint>
      <joint name="right_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>  
      </joint>
    </ros2_control>
  </xacro:unless>

  <xacro:if value="$(arg sim_mode)">
    <ros2_control name="GazeboSimSystem" type="system">
      <hardware>
        <plugin>gz_ros2_control/GazeboSimSystem</plugin>
      </hardware>
      <joint name="left_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="velocity" />
        <state_interface name="position" />
      </joint>
      <joint name="right_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="velocity" />
        <state_interface name="position" />
      </joint>
    </ros2_control>
  </xacro:if>

  <gazebo>
    <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
      <ros>
        <namespace>$(arg namespace)</namespace>
        <remapping>/diff_drive_base_controller/cmd_vel_unstamped:=/cmd_vel</remapping>
      </ros>
      <parameters>$(find 2025_esda_simulation)/config/my_controllers_2wd.yaml</parameters>
      <parameters>$(find 2025_esda_simulation)/config/gaz_ros2_ctl_use_sim.yaml</parameters>
    </plugin>
  </gazebo>

</robot>